### ISSUE TITLE
Ensure consistent Chrome versions during CI-parallelization

### ISSUE THREAD

**Issue by muratkeremozcan** [cypress-io/cypress#25040]
### Current behavior

Here's a very ambiguous CI issue during parallelization

It happens because Chrome rolls out varying versions , and if the parallelized CI machines do not match versions, you get the error.

Sporadic issue. It's the first time I saw it externally.

https://github.com/muratkeremozcan/tour-of-heroes-react-cypress-ts/actions/runs/3642676076/jobs/6150068190 

```
You passed the --parallel flag, but we do not parallelize tests across different environments.

This machine is sending different environment parameters than the first machine that started this parallel run.

The existing run is: https://cloud.cypress.io/projects/7mypio/runs

In order to run in parallel mode each machine must send identical environment parameters such as:

 - specs
 - osName
 - osVersion
 - browserName
 - browserVersion (major)

This machine sent the following parameters:

{
  "osName": "linux",
  "osVersion": "Ubuntu - ",
  "browserName": "Chrome",
  "browserVersion": "107.0.5304.121",
```

To work around it, we have to ensure the same Chrome version in all CI machines

https://github.com/muratkeremozcan/tour-of-heroes-react-cypress-ts/pull/187/commits/f0d37f6fd9bea8d124a30135d2193a8d6c04abc0

```
      - name: Install Specific Chrome Version
        run: |
          sudo apt-get install -y wget
          sudo wget -q https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
          sudo apt-get install ./google-chrome-stable_current_amd64.deb

```

This adds CI minutes. It would be ideal if Cypress had a solution to it for parallelization.

### Desired behavior

It would be ideal if Cypress Cloud, or the GHA handled parallelization and ensuring consistent Chrome versions between CI machines

### Test code to reproduce

See links in description. You will not reproduce this one with code. It will happen sporadically the more parallel CI machines you have.

### Cypress Version

12.0.1

### Node version

lts - whatever Cypress GHA uses

### Operating System

CI - ubuntu

### Debug Logs

```shell
See description.
```

**Comment by muratkeremozcan** [comment]
From @conversaShawn 

"Thanks for reporting Murat. The current workaround is to use a specific version as you did or use one of Cypress images: https://github.com/cypress-io/cypress-docker-images/tree/master/browsers
Currently, this typically happens whenever a new browser version is released since it grabs the latest. I've also seen the happen when using ubuntu-latest 😢 
The good news is, we are actively investigating, so I'll loop in your CSM to keep you in the loop of any updates"

**Comment by conversayShawn** [comment]
Related to cypress/cypress issues:
- [#7199](https://github.com/cypress-io/cypress/issues/7199#issuecomment-1143628102)

Related to cypress/github-action issues:
- [121](https://github.com/cypress-io/github-action/issues/121)
- [602](https://github.com/cypress-io/github-action/issues/602)


Existing workaround:
- [Comment](https://github.com/cypress-io/cypress/issues/7199#issuecomment-1143628102)

**Comment by a28jiang** [comment]
Would ideally like to avoid the current workaround as it adds additional overhead to our job duration. It would be great if`parallel-group-params-mismatch` could be made less strict!

**Comment by essenmitsosse** [comment]
Since Chrome 111 was just released, I am encountering this error again. I also agree that I'd rather have `parallel-group-params-mismatch` less strict if there isn't a more elegant version to ensure all CIs use the same browser. But this needs to be solved. Every few weeks, Cypress tests get super brittle.

If the solution is manually downloading a pinned image, which would add significant time to CI, I'd much rather turn off parallelization and save money on Cypress Cloud. What I want to say: This bug is actively hurting your paid product and making it partially useless.

**Comment by jmtimko5** [comment]
We are also see this problem. Especially since default versions of browsers on Github runners are not under our control, it would be great if via configurability we could say certain parameters are no of concern.

**Comment by arya** [comment]
We’re also running into this and don’t want to add significant time to our CI and don’t see a viable workaround. Is there one?

Maybe one option for the cypress team is to add a flag (or just always) to change the behavior to warn on browser version differences. The warning can be close to the test errors, so that if browser versions cause test failures it can be apparent why.

**Comment by essenmitsosse** [comment]
@ayra: Funny enough, automatic updating (which could cause test errors) isn't even the issue here — it's only when they are out of sync because some workers auto-updated and others didn't. So most users wouldn't even want to be warned about version mismatched. If a new version actually causes a test failure, you are dealing with a whole different can of worms.

**Comment by arya** [comment]
What’s the reason to not attempt running the tests when the versions are out of sync? It’s not clear why not trying to run the tests when one is on Chrome 110 and one is one Chrome 111

**Comment by nirmpate** [comment]
In the end we decided to switch our tests to run on electron for the short term just to unblock the team from merging PR's.  This worked fine and didn't slow down our pipeline.  We're thinking of switching back to Chrome after a couple days.  The cypress team should find a real solution to this or else we're stilling going to be getting the same problem once v112 comes out.

**Comment by essenmitsosse** [comment]
Chrome 112 is out, and it is happening again. Very disappointed about the radio silence because this will cause us a lot of hassle again, and parallelization is the only reason we are paying for Cypress. If this keeps breaking every few weeks, we'd rather switch to the free version and turn off parallelization.

I want to point out this is a core feature of the paid product, and it reliably breaks once a month for like a week.

**Comment by MikeMcC399** [comment]
Adding one of the `cypress/browsers` Docker containers to the workflow should work around the issue that GitHub is jumping backwards and forwards in its use of image versions in the same job and different matrix containers during a rollout period.

Docker lists the available containers on https://hub.docker.com/r/cypress/browsers/tags

The GitHub Action CI documentation for [Parallelization](https://docs.cypress.io/guides/continuous-integration/github-actions#Parallelization) shows the use of a Docker container, although the current example `container: cypress/browsers:node12.18.3-chrome87-ff82` is outdated.

One example would be:

```
jobs:
  install:
    runs-on: ubuntu-22.04
    container: cypress/browsers:node18.12.0-chrome107
```

otherwise there are newer images which have Chrome, Firefox and Edge included.

Edit: GitHub Actions product feedback added in https://github.com/orgs/community/discussions/52956.

**Comment by MikeMcC399** [comment]
[cypress-io/github-action > README > Parallel](https://github.com/cypress-io/github-action/blob/master/README.md#parallel) now contains the specific advice:

"During staged rollout of a new GitHub-hosted runner version, GitHub may provide a mixture of current and new image versions used by the container matrix. It is recommended to use a [Docker image](https://github.com/cypress-io/github-action/blob/master/README.md#docker-image) in the parallel job run which avoids any Cypress Cloud errors due to browser major version mismatch from the two different image versions. A [Docker image](https://github.com/cypress-io/github-action/blob/master/README.md#docker-image) is not necessary if testing against the default built-in Electron browser because this browser version is fixed by the Cypress version in use and it is unaffected by any GitHub runner image rollout."

**Comment by MikeMcC399** [comment]
Cypress documentation
[Specifying Browsers in Parallel Builds](https://docs.cypress.io/guides/continuous-integration/github-actions#Specifying-Browsers-in-Parallel-Builds) now describes precautions to prevent this issue causing errors:

![image](https://github.com/cypress-io/cypress/assets/66998419/ae3ff03d-ca02-4122-903e-f18bc8afcdf6)

GitHub runners are currently preparing to update from Google Chrome `112.0.5615.165` to `113.0.5672.63`.

GitHub product management didn't respond so far to https://github.com/orgs/community/discussions/52956

**Comment by MikeMcC399** [comment]
The workaround of using a [Cypress Docker image](https://docs.cypress.io/guides/continuous-integration/github-actions#Testing-with-Cypress-Docker-Images) is only available for Ubuntu / Linux on GitHub. These are the only operating systems supported as containers on GitHub. Windows and macOS are not supported. See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer

![image](https://github.com/cypress-io/cypress/assets/66998419/2da7a27a-e346-418b-8754-6276f4b606ee)

**Comment by MikeMcC399** [comment]
- According to the feedback in the GitHub actions/runner issue https://github.com/actions/runner/issues/2812#issuecomment-1708710610 no change can be expected in the way that GitHub Actions Backend randomly assigns image versions in response to requests for a virtual machine. So the current best advice is still to use a Docker image or otherwise install fixed browser versions when recording to Cypress Cloud.

**Comment by nick-fields** [comment]
Our app requires MacOS and bare metal GPU access, so the docker images are not helpful here. What compounds the issue is Chrome's rolling updates, and how they don't really support disabling auto updates (the recommended Keystone plist updates seem to be no longer supported).

Can the check for browserVersion at least be made configurable?

**Comment by a8trejo** [comment]
Don't know why i don't see it highlighted here, but this workaround has worked for me @nick-fields 
https://github.com/cypress-io/github-action/issues/518#issuecomment-1210979047

**Comment by bmulholland** [comment]
I'm hitting this error now and it just seems so strange. Cypress has intentionally added a conditional that crashes some workers during a run, which breaks builds... but why? Has anyone received an explanation for why this is the case? I'm fine with Chrome being different between different workers on the same runs. This effectively breaks the parallel test runs that we pay for with Cypress, doesn't it? Am I missing something?

I believe any workaround for this requires us to manually manage Chrome versions, monitoring Chrome updates and changing a version number manually, and/or additional CI minutes to install versions. Is there one that doesn't cost me CI time and doesn't require us to add yet another version of something to manage and upgrade over time?

**Comment by jennifer-shehane** [comment]
The logic errors when there is a difference in MAJOR browser version between two specs within the same parallelized run. A core component of testing and getting reliable results is ensuring a stable environment in which all of the tests are run. If one specfile runs in Chrome 119 and one runs in Chrome 120, you may get mixed results due to the vast amount of changes that Chrome makes in major updates. We encountered this ourselves - our own internal tests at Cypress began failing with a Chrome update when they changed their behavior around http to https upgrades. You would end up with mixed results in this situation.

I would encourage you to look towards solutions that maintain a stable browser version to run your tests within - using our provided [docker-images](https://github.com/cypress-io/cypress-docker-images) or other solutions that ensure browsers don't auto-update in the middle of runs. 

We'll leave this issue open for future consideration, but to accomplish this would be Cypress providing our own browsers and a versioning system which we spent some time investigating but is a lot of work to implement.

**Comment by bmulholland** [comment]
Thanks for the response, @jennifer-shehane. I understand that upgrades like that may cause inconsistencies in results -- and I'm happy to take on the responsibility for that. I'd rather it over Cypress failing my test suite entirely, or having to manage Chrome versions manually. Why doesn't Cypress give me an option to "enable browser inconsistencies" or something? Opt-in, with a clear warning, and just disable this one check.

**Comment by JoseAlban** [comment]
I have a situation on Github CI in which I install the browser via:
```
      - uses: browser-actions/setup-chrome@v1 # default to latest
        id: setup-chrome
        with:
          chrome-version: stable
```
which installed the same chrome version on all 8 of 8 parallel jobs
```
Successfully setup chromium 128.0.6613.84
```

then set the browser via
```
      - name: Cypress run
        uses: cypress-io/github-action@v6
        with:
          browser: ${{ steps.setup-chrome.outputs.chrome-path }}
```

using the ubuntu 24 runner
```
runs-on: ubuntu-24.04
```

and cypress parallel failed me, with a list of browsers of
```
{
      name: 'chrome',
      family: 'chromium',
      channel: 'stable',
      displayName: 'Chrome',
      version: '127.0.6533.119',
      path: 'google-chrome',
      minSupportedVersion: 64,
      majorVersion: '127'
    },
    {
      name: 'chrome',
      family: 'chromium',
      channel: 'stable',
      displayName: 'Chrome',
      version: '128.0.6613.84',
      path: 'chrome',
      minSupportedVersion: 64,
      majorVersion: '128'
    },
    {
      name: 'chromium',
      family: 'chromium',
      channel: 'stable',
      displayName: 'Chromium',
      version: '127.0.6533.0',
      path: 'chromium-browser',
      minSupportedVersion: 64,
      majorVersion: '127'
    },
    {
      name: 'electron',
      channel: 'stable',
      family: 'chromium',
      displayName: 'Electron',
      version: '118.0.5993.159',
      path: '',
      majorVersion: 118
    }
```
and error
```
  "browserVersion": "128.0.6613.84.... (Expected: 127)",
```

so, it sounds quite dumb that cypress is looking up the major version in a way that's not related to the specified `browser` parameter's bin version.

**Comment by MikeMcC399** [comment]
@JoseAlban 

According to the [GitHub Runners dashboard](https://github.com/actions/runner-images) there is currently no active deployment for the beta runner [Ubuntu 24.04](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md) so you have hit a different issue from the one described by the original post here.

The beta version [Ubuntu 24.04](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#browsers-and-drivers) is currently deployed with

- Google Chrome 127.0.6533.119
- ChromeDriver 127.0.6533.119
- Chromium 127.0.6533.0
- Selenium server 4.23.1

You write that you added Chrome `stable` / Chromium `128.0.6613.84` via [browser-actions/setup-chrome](https://github.com/browser-actions/setup-chrome).

Your error is:

>  "browserVersion": "128.0.6613.84.... (Expected: 127)",

This would need investigation to find out whether the problem is coming from Cypress, Cypress Cloud or `cypress-io/github-action`. I suspect it is a Cypress Cloud error, but that is only a guess and it would need logs to find out.

Chrome `128.0.6613.84` (Linux) was [released yesterday](https://chromereleases.googleblog.com/2024/08/stable-channel-update-for-desktop_21.html) Aug 21, 2024 and I would expect that the next release of Cypress Docker images will pick up this version, so you would probably be able to access Chrome `128.x` alternatively with a `cypress/browsers` image in a few days time. Edit: You can also create a custom Cypress Docker image at any time, with your own choice of browsers, using [cypress/factory](https://github.com/cypress-io/cypress-docker-images/tree/master/factory).

**Comment by beaugunderson** [comment]
hmm... this is breaking again for us.

<img width="455" height="170" alt="Image" src="https://github.com/user-attachments/assets/c53bc1ea-ec9f-4c33-9549-6bf529ac63c8" />

we're specifying a fixed version for Chrome of `140.0.7339.80` but apparently some runners are sending `140`?

**Comment by MikeMcC399** [comment]
@beaugunderson 

> hmm... this is breaking again for us.
> 
> <img alt="Image" width="455" height="170" src="https://private-user-images.githubusercontent.com/61791/487403455-c53bc1ea-ec9f-4c33-9549-6bf529ac63c8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc0ODY1MzEsIm5iZiI6MTc1NzQ4NjIzMSwicGF0aCI6Ii82MTc5MS80ODc0MDM0NTUtYzUzYmMxZWEtZWM5Zi00YzMzLTk1NDktNmJmNTI5YWM2M2M4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTEwVDA2MzcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNTJlNDM2OGJjNjk5YWM1YjM4NmVjYTcwMWYyNDhhYWUzMGUwMTA1N2EwYmZhMDVlODMwZDIxMGJlNWUzZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.t5fmgwy5xi4PCZ46CgwHpFai6yl8XA7Gz5OAXRzs03g">
> we're specifying a fixed version for Chrome of `140.0.7339.80` but apparently some runners are sending `140`?

For Linux systems, Cypress recommends using the Cypress Docker image `cypress/browsers`. See https://docs.cypress.io/app/continuous-integration/github-actions#Specifying-Browsers-in-Parallel-Builds.

**Comment by MikeMcC399** [comment]
@beaugunderson 

To add to and replace some of what I wrote in the previous comment ⏫ 

> "BrowserVersion": "140.0.7339.80 ... (Expected: 140)",

I would not expect to see that error message since `140.0.7339.80` already matches the major version `140`. According to the explanation in https://github.com/cypress-io/cypress/issues/25040#issuecomment-1866570859 that should not be causing an error.

This may be a bug in Cypress Cloud.

If this is something that you would like to investigate further, I suggest opening a new issue here to gather further details, like how you are setting up the version of Chrome to use for testing. If you are using `cypress-io/github-action`, you can instead open a new issue in https://github.com/cypress-io/github-action/issues. (Cypress Cloud also has its own separate support channel - see https://www.cypress.io/support).

As an aside:

> "osVersion": "Ubuntu - ",

[Cypress 15.1.0](https://docs.cypress.io/app/references/changelog#15-1-0) fixes issue https://github.com/cypress-io/cypress/issues/30533 where the Linux version (like `24.04`) was not reported in Cypress Cloud.

**Comment by MikeMcC399** [comment]
@jennifer-shehane 

> We'll leave this issue open for future consideration, but to accomplish this would be Cypress providing our own browsers and a versioning system which we spent some time investigating but is a lot of work to implement.

I'm not sure why this issue would need to be kept open anymore.

In a GitHub Actions matrix environment, there are two different solutions to lock browsers across parallel jobs without depending on the random browser versions that GitHub runner images provide during a runner image upgrade transition phase:

- use a Docker image `cypress/browsers` 
- use one of the GitHub Actions from [browser-actions](https://github.com/browser-actions) such as is described in the [cypress-io/github-action README for Chrome for Testing](https://github.com/cypress-io/github-action/blob/master/README.md#chrome-for-testing)

If there are any issues using these, they can be handled in separate individual issues. I did attempt to simulate the issue that @beaugunderson reported, however it worked as expected. If they can provide some details, we can follow this up.

**Comment by jennifer-shehane** [comment]
Thanks for sharing everyone. We’re tracking this feedback on the Product team.

I will close this issue on the public repo since this is technically Cypress Cloud logic, which is reserved for App issues/features.
